pkg/cmd/readyz/readyz.go

```diff
@@ -0,0 +1,189 @@
+	// https://github.com/kubernetes/apiserver/blob/de6ba2aa0a752d077719fcee186291d7af73a825/pkg/storage/storagebackend/factory/etcd3.go#L55-L56
+	keepaliveTime     = 30 * time.Second
+	keepaliveTimeout  = 10 * time.Second
+	defaultListenPort = 9980
+)
+
+type readyzOpts struct {
+	listenPort       uint16
+	dialTimeout      time.Duration
+	targetEndpoint   string
+	clientCertFile   string
+	clientKeyFile    string
+	clientCACertFile string
+}
+
+// NewReadyzCommand creates a readyz command that runs as an http-get readiness server alongside the etcd member container
+func NewReadyzCommand() *cobra.Command {
+	opts := readyzOpts{
+		listenPort: uint16(defaultListenPort),
+	}
+	cmd := &cobra.Command{
+		Use:   "readyz",
+		Short: "Serve the HTTP /readyz endpoint health check for an etcd member",
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := opts.Validate(); err != nil {
+				klog.Fatal(err)
+			}
+			if err := opts.Run(); err != nil {
+				klog.Fatal(err)
+			}
+		},
+	}
+
+	opts.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (r *readyzOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.Uint16Var(&r.listenPort, "listen-port", r.listenPort, "Listen on this port. Default 9980")
+	fs.DurationVar(&r.dialTimeout, "dial-timeout", defaultHTTPDialTimeout, "Dial timeout for the client. Default 2s")
+	fs.StringVar(&r.targetEndpoint, "target", defaultEndpoint, "Target endpoint to perform health check against. Default https://localhost:2379")
+	fs.StringVar(&r.clientCertFile, "cert-file", "", "Health probe TLS client certificate file. (required)")
+	fs.StringVar(&r.clientKeyFile, "key-file", "", "Health probe TLS client key file. (required)")
+	fs.StringVar(&r.clientCACertFile, "cacert-file", "", "Health probe TLS client CA certificate file. (required)")
+}
+
+// Validate verifies the inputs.
+func (r *readyzOpts) Validate() error {
+	if len(r.targetEndpoint) == 0 {
+		return fmt.Errorf("missing required flag: --target")
+	}
+	if len(r.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(r.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(r.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+
+	return nil
+}
+
+// Run contains the logic of the readyz command which checks the health of the etcd member
+func (r *readyzOpts) Run() error {
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	shutdownHandler := server.SetupSignalHandler()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/readyz", r.getReadyzHandlerFunc(shutdownCtx))
+	// Handle the /healthz endpoint as well since the static pod controller's guard pods check the /healthz endpoint
+	// https://github.com/openshift/library-go/blob/edab248e63516c65a93467eaa8224c86d69f5de9/pkg/operator/staticpod/controller/guard/manifests/guard-pod.yaml#L44
+	mux.HandleFunc("/healthz", r.getReadyzHandlerFunc(shutdownCtx))
+
+	addr := fmt.Sprintf("0.0.0.0:%d", r.listenPort)
```

Comment:
can we bind to localhost?

Reply:
I tested it out but I'm not sure that works.
The http-get probe on the `etcd` container will pass if I additionally specify `host: 127.0.0.1` since we are using `hostNetwork: true`.
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes

But on the actual guard pod the probe uses the etcd pod's IP as the host which isn't accessible since the readyz server is listening on localhost only. So the guard pod's probe fails to reach the readyz server on the etcd pod.
```YAML
      readinessProbe:
        httpGet:
          path: healthz
          port: 9980
          host: 10.0.223.164
          scheme: HTTPS
```
```console
$ kubectl -n openshift-etcd describe pod etcd-guard-ip-10-0-223-164.us-west-1.compute.internal
Events:
  Type     Reason          Age                    From     Message
  ----     ------          ----                   ----     -------
...
  Normal   Started         4m5s                   kubelet  Started container guard
  Warning  ProbeError      3m23s (x11 over 4m4s)  kubelet  Readiness probe error: Get "https://10.0.223.164:9980/healthz": dial tcp 10.0.223.164:9980: connect: connection refused
body:
  Warning  Unhealthy  3m23s (x11 over 4m4s)  kubelet  Readiness probe failed: Get "https://10.0.223.164:9980/healthz": dial tcp 10.0.223.164:9980: connect: connection refused
```

Reply:
Ok, thanks